### PR TITLE
Mepts 1076 tpt initiation changes

### PR DESCRIPTION
**Pull request by TinoMuianga**

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationDataDefinitionQueries.java [hunk 1]
```diff
@@ -1162,241 +1160,245 @@ public DataDefinition getDifferencebetweenRegisteredvsExpected3HPCompletionDate(
...
+            + "                     AND o.value_coded IN ( ${23954}  ,  ${23984} ) "
+            + "                     AND e.encounter_datetime BETWEEN :startDate AND :endDate "
+            + "                     GROUP BY p.patient_id "
+            + "       UNION "
+            + "              SELECT p.patient_id, MIN(o2.obs_datetime) AS consultation_date "
+            + "              FROM patient p "
+            + "                    INNER JOIN encounter e ON p.patient_id = e.patient_id "
+            + "                    INNER JOIN obs o ON e.encounter_id = o.encounter_id "
+            + "                    INNER JOIN obs o2 ON e.encounter_id = o2.encounter_id "
+            + "                    INNER JOIN obs o3 ON e.encounter_id = o3.encounter_id "
```
Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationDataDefinitionQueries.java]: I think this o3 is not being used!

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationDataDefinitionQueries.java [hunk 2]
```diff
@@ -1162,241 +1160,245 @@ public DataDefinition getDifferencebetweenRegisteredvsExpected3HPCompletionDate(
...
+            + "    SELECT p.patient_id "
+            + "        FROM patient p "
+            + "            INNER JOIN encounter e ON p.patient_id = e.patient_id "
+            + "            INNER JOIN obs o ON e.encounter_id = o.encounter_id "
+            + "            INNER JOIN obs o2 ON e.encounter_id = o2.encounter_id "
+            + "        WHERE p.voided = 0 AND e.voided = 0 AND o.voided = 0 AND o2.voided = 0 "
+            + "            AND e.location_id = :location "
+            + "            AND e.encounter_type = ${60} "
+            + "            AND ( ( o.concept_id = ${23985} AND o.value_coded IN ( ${23954}, ${23984} ) ) "
+            + "            AND ( o2.concept_id = ${23987} AND o2.value_coded IN ( ${1257}, ${1267} ) OR o2.value_coded IS NULL "
```
Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationDataDefinitionQueries.java]:
```suggestion
            + "            AND ( o2.concept_id = ${23987} AND (o2.value_coded IN ( ${1257}, ${1267} ) OR o2.value_coded IS NULL) "
```

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationCohortQueries.java [hunk 3]
```diff
@@ -389,7 +404,7 @@ public CohortDefinition getPatientsWithRegimeDeTPT3HP() {
             + "       AND e.encounter_type = ${60} "
             + "       AND ( ( o.concept_id = ${23985} AND o.value_coded IN ( ${23954}, ${23984} ) ) "
             + "       AND ( o2.concept_id = ${23987} AND o2.value_coded IN ( ${1257}, ${1267} ) OR o2.value_coded IS NULL "
```
Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationCohortQueries.java]:
```suggestion
            + "       AND ( o2.concept_id = ${23987} AND (o2.value_coded IN ( ${1257}, ${1267} ) OR o2.value_coded IS NULL) "
```

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationCohortQueries.java [hunk 4]
```diff
@@ -695,20 +700,20 @@ public CohortDefinition getPatientsWithFirstFiltRegimeTpt() {
...
             + "                                    FROM patient p INNER JOIN encounter e ON e.patient_id = p.patient_id "
             + "               INNER JOIN obs o ON o.encounter_id = e.encounter_id "
             + "               INNER JOIN obs o2 on o2.encounter_id = e.encounter_id "
             + "                                    WHERE   p.voided = 0  AND e.voided = 0 "
             + "               AND o.voided = 0  AND e.location_id = :location "
             + "               AND e.encounter_type = ${60} "
-            + "               AND ( o.concept_id = ${23985}  AND o.value_coded IN (${656}, ${23982}) ) "
-            + "               AND ((o2.concept_id =  ${23987}  AND (o2.value_coded =  ${1257}  OR o2.value_coded IS NULL)) "
+            + "               AND ( ( o.concept_id = ${23985}  AND o.value_coded IN (${656}, ${23982}) ) "
+            + "               AND   ( o2.concept_id =  ${23987}  AND (o2.value_coded =  ${1257}  OR o2.value_coded IS NULL) "
```
Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationCohortQueries.java]:
```suggestion
            + "               AND   ( (o2.concept_id =  ${23987}  AND (o2.value_coded =  ${1257}  OR o2.value_coded IS NULL)) "
```

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationDataDefinitionQueries.java [hunk 5]
```diff
@@ -949,9 +949,9 @@ public DataDefinition getExpected3HPCompletionDate() {
             + "        WHERE p.voided = 0 AND e.voided = 0 AND o.voided = 0 AND o2.voided = 0 "
             + "            AND e.location_id = :location "
             + "            AND e.encounter_type = ${60} "
-            + "            AND ( o.concept_id = ${23985} AND o.value_coded IN ( ${23954}, ${23984} ) ) "
-            + "            AND ( o2.concept_id = ${23987} AND o2.value_coded IN ( ${1257}, ${1267} ) OR o2.value_coded IS NULL ) "
-            + "            AND e.encounter_datetime BETWEEN :startDate AND :endDate "
+            + "            AND ( ( o.concept_id = ${23985} AND o.value_coded IN ( ${23954}, ${23984} ) ) "
+            + "            AND ( o2.concept_id = ${23987} AND o2.value_coded IN ( ${1257}, ${1267} ) OR o2.value_coded IS NULL "
```
Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationDataDefinitionQueries.java]:
```suggestion
            + "            AND ( o2.concept_id = ${23987} AND (o2.value_coded IN ( ${1257}, ${1267} ) OR o2.value_coded IS NULL) "
```

**Review by psmatsinhe — CHANGES_REQUESTED** [review]
@TinoMuianga please have a look on TB3 review comments

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationDataDefinitionQueries.java (hunk 1 above)
Comment (TinoMuianga) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationDataDefinitionQueries.java]: Resolved

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationDataDefinitionQueries.java (hunk 2 above)
Comment (TinoMuianga) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationDataDefinitionQueries.java]: Resolved

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationCohortQueries.java (hunk 3 above)
Comment (TinoMuianga) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationCohortQueries.java]: Resolved

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationCohortQueries.java (hunk 4 above)
Comment (TinoMuianga) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationCohortQueries.java]: Resolved

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationDataDefinitionQueries.java [hunk 6]
```diff
@@ -1162,241 +1160,244 @@ public DataDefinition getDifferencebetweenRegisteredvsExpected3HPCompletionDate(
...
+            + "        FROM patient p "
+            + "            INNER JOIN encounter e ON p.patient_id = e.patient_id "
+            + "            INNER JOIN obs o ON e.encounter_id = o.encounter_id "
+            + "            INNER JOIN obs o2 ON e.encounter_id = o2.encounter_id "
+            + "        WHERE p.voided = 0 AND e.voided = 0 AND o.voided = 0 AND o2.voided = 0 "
+            + "            AND e.location_id = :location "
+            + "            AND e.encounter_type = ${60} "
+            + "            AND ( ( o.concept_id = ${23985} AND o.value_coded IN ( ${23954}, ${23984} ) ) "
+            + "            AND ( o2.concept_id = ${23987} AND o2.value_coded IN ( ${1257}, ${1267} ) OR o2.value_coded IS NULL ) "
+            + "            AND o2.obs_datetime BETWEEN :startDate AND :endDate ) "
```
Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationDataDefinitionQueries.java]:
```suggestion
            + "            AND ( o2.concept_id = ${23987} AND ( o2.value_coded IN ( ${1257}, ${1267} ) OR o2.value_coded IS NULL ) "
            + "            AND o2.obs_datetime BETWEEN :startDate AND :endDate ) ) "
```

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationDataDefinitionQueries.java [hunk 7]
```diff
@@ -949,9 +949,9 @@ public DataDefinition getExpected3HPCompletionDate() {
             + "        WHERE p.voided = 0 AND e.voided = 0 AND o.voided = 0 AND o2.voided = 0 "
             + "            AND e.location_id = :location "
             + "            AND e.encounter_type = ${60} "
-            + "            AND ( o.concept_id = ${23985} AND o.value_coded IN ( ${23954}, ${23984} ) ) "
+            + "            AND ( ( o.concept_id = ${23985} AND o.value_coded IN ( ${23954}, ${23984} ) ) "
             + "            AND ( o2.concept_id = ${23987} AND o2.value_coded IN ( ${1257}, ${1267} ) OR o2.value_coded IS NULL ) "
```
Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationDataDefinitionQueries.java]:
```suggestion
            + "            AND ( o2.concept_id = ${23987} AND ( o2.value_coded IN ( ${1257}, ${1267} ) OR o2.value_coded IS NULL ) "
```

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationDataDefinitionQueries.java [hunk 8]
```diff
@@ -949,9 +949,9 @@ public DataDefinition getExpected3HPCompletionDate() {
...
-            + "            AND e.encounter_datetime BETWEEN :startDate AND :endDate "
+            + "            AND o2.obs_datetime BETWEEN :startDate AND :endDate ) "
```
Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationDataDefinitionQueries.java]:
```suggestion
            + "            AND o2.obs_datetime BETWEEN :startDate AND :endDate )) "
```

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationCohortQueries.java [hunk 9]
```diff
@@ -388,8 +403,8 @@ public CohortDefinition getPatientsWithRegimeDeTPT3HP() {
             + "       AND e.location_id = :location "
             + "       AND e.encounter_type = ${60} "
             + "       AND ( ( o.concept_id = ${23985} AND o.value_coded IN ( ${23954}, ${23984} ) ) "
-            + "       AND ( o2.concept_id = ${23987} AND o2.value_coded IN ( ${1257}, ${1267} ) OR o2.value_coded IS NULL "
-            + "       AND o2.obs_datetime BETWEEN DATE_SUB(:endDate, INTERVAL 4 MONTH) AND :endDate ) ) "
+            + "       AND ( o2.concept_id = ${23987} AND o2.value_coded IN ( ${1257}, ${1267} ) OR o2.value_coded IS NULL ) "
+            + "       AND o2.obs_datetime BETWEEN :startDate AND :endDate ) "
```
Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationCohortQueries.java]:
```suggestion
            + "       AND ( o2.concept_id = ${23987} AND ( o2.value_coded IN ( ${1257}, ${1267} ) OR o2.value_coded IS NULL ) "
            + "       AND o2.obs_datetime BETWEEN :startDate AND :endDate )) "
```

**Review by psmatsinhe — CHANGES_REQUESTED** [review]
@TinoMuianga Please have a look on this

**Review by psmatsinhe — APPROVED** [review]
Looks good to me

**Pull request merged**